src/appengine/handlers/base_handler_flask.py

```diff
@@ -0,0 +1,303 @@
+    values['is_oss_fuzz'] = utils.is_oss_fuzz()
+    values['is_development'] = (
+        environment.is_running_on_app_engine_development())
+    values['is_logged_in'] = bool(helpers.get_user_email())
+
+    # Only track analytics for non-admin users.
+    values['ga_tracking_id'] = (
+        local_config.GAEConfig().get('ga_tracking_id')
+        if not auth.is_current_user_admin() else None)
+
+    if values['is_logged_in']:
+      values['switch_account_url'] = make_login_url(request.url)
+      values['logout_url'] = make_logout_url(dest_url=request.url)
+
+    template = _JINJA_ENVIRONMENT.get_template(path)
+
+    response = Response()
+    response = self._add_security_response_headers(response)
+    response.headers['Content-Type'] = 'text/html'
+    response.data = template.render(values)
+    response.status_code = status
+    return response
+
+  # pylint: disable=unused-argument
+  def before_render_json(self, values, status, response):
+    """A hook for modifying values before render_json."""
+    return response
+
+  def render_json(self, values, status=200):
+    """Write JSON response."""
+    response = Response()
+    response = self._add_security_response_headers(response)
+    response.headers['Content-Type'] = 'application/json'
+    response = self.before_render_json(values, status, response)
+    response.data = json.dumps(values, cls=JsonEncoder)
+    response.status_code = status
+    return response
+
+  def handle_exception(self, exception):
+    """Catch exception and format it properly."""
+    try:
+      status = 500
+      values = {
+          'message': str(exception),
+          'email': helpers.get_user_email(),
+          'traceDump': traceback.format_exc(),
+          'status': status,
+          'type': exception.__class__.__name__
+      }
+      if isinstance(exception, helpers.EarlyExitException):
+        status = exception.status
+        values = exception.to_dict()
+
+      # 4XX is not our fault. Therefore, we hide the trace dump and log on
+      # the INFO level.
+      if 400 <= status <= 499:
+        logging.info(json.dumps(values, cls=JsonEncoder))
+        del values['traceDump']
+      else:  # Other error codes should be logged with the EXCEPTION level.
+        logging.exception(exception)
+
+      if self.is_json or ('application/json' in request.headers.get(
+          'accept', '')):
+        return self.render_json(values, status)
+      if status in (403, 401):
+        return self.render_forbidden(str(exception))
+      return self.render('error.html', values, status)
+    except Exception:
+      self.handle_exception_exception()
+
+  def handle_exception_exception(self):
+    """Catch exception in handle_exception and format it properly."""
+    exception = sys.exc_info()[1]
+    values = {'message': str(exception), 'traceDump': traceback.format_exc()}
+    logging.exception(exception)
+    if self.is_json or ('application/json' in request.headers.get(
```

Comment:
actually, since this is repeated, can just put this into a method in this class.